### DISCUSSION TITLE
Generate prettier grammar from ANTLR or Atom grammar

### DISCUSSION THREAD

**Discussion by thorade** [prettier/prettier#16901] [Q&A] [answered]
Is anybody aware of a tool or script that can generate a first version of a prettier grammar (or language plugin) from an existing ANTLR,  treesitter or Atom/Sublime/VScode grammar?

- https://specification.modelica.org/maint/3.6/lexical-structure.html
- https://github.com/SimplyDanny/modelica-language-vscode
- https://borischumichev.github.io/modelicaSublimeTextPackage/
- https://github.com/modelica-tools/modelica-mode
- https://github.com/modelica/ModelicaSpecification/pull/2378
- https://github.com/OpenModelica/tree-sitter-modelica/issues/25

Other Modelica formatting tools mentioned here:
https://stackoverflow.com/questions/14163386/modelica-style-guide

**Comment by Poliklot** [comment, marked as answer]
A grammar can help with the parser part, but it will not get you a usable Prettier plugin by itself.

ANTLR or tree-sitter can generate the parser part, but that is only one half of a Prettier plugin. The hard part is the printer: converting the AST/CST into Prettier's Doc IR and deciding how every language construct should be laid out.

TextMate / Atom / Sublime / VSCode grammars are even further away from this use case: they are mainly syntax-highlighting grammars, not formatter-ready parsers.

For Modelica, a realistic path would be:

1. reuse an existing ANTLR or tree-sitter grammar to parse the language;
2. decide whether to print directly from the CST or normalize it into a smaller AST;
3. implement the Prettier printer manually;
4. add snapshot tests and grow coverage construct by construct.

So a grammar can give you a starting parser, but it will not generate the formatting rules or the Prettier printer for you.

**Reply by thorade** [reply]
Thanks for that explanation!

**Comment by thorade** [comment]
In case someone finds this question looking for a Modelica prettier grammar, @AntoineGautier published one here:
https://github.com/AntoineGautier/prettier-plugin-modelica
